<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />


<title>Using BLAST on the command line</title>

<script src="blast_protein-family_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="blast_protein-family_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="blast_protein-family_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="blast_protein-family_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="blast_protein-family_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="blast_protein-family_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="blast_protein-family_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="blast_protein-family_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="blast_protein-family_files/navigation-1.1/tabsets.js"></script>
<script src="blast_protein-family_files/navigation-1.1/codefolding.js"></script>




<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Using BLAST on the command line</h1>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">Created: 8/10/2018; last update: 2019-11-28</h4>

</div>


<hr />
<div id="goals-of-this-tutorial" class="section level1">
<h1>Goals of this tutorial</h1>
<p>In this tutorial, we will use the local sequence similarity search tool (BLAST) to searching sequence similarities for a protein of interest (an histone) in the proteome of a given organism (the yeast Saccharomyces cerevisiae).</p>
<p>This approach is very simple, fast and widely used (BLAST is the most highly bioinformatics tool worldwide), but suffers from some limitations. In particular, the resulting list of protein sequences is sensitive to the particular protein chosse as query.</p>
<p>In a next practical, we will rely on a more powerful method based on Hidden Markov Models (HMM) in order to build a profile for a protein family, and use this profile to scan the proteome and detect all the members of this family.</p>
<p>Technically, this tutorial will also lead us to submit jobs to a scheduler, monitor the output and error logs, and organise the results for further use.</p>
<p>We will also perform some exercise to evaluate the impact of some parameters on the results (in particular, the threshold on the e-value).</p>
<hr />
</div>
<div id="operations" class="section level1">
<h1>Operations</h1>
<p>We will successively perform the following operations.</p>
<ol style="list-style-type: decimal">
<li><p>Select the full set of proteins (proteome) identified by translating the putatively coding regions of a given organism (e.g. the fruit fly <em>Drosophila melanogaster</em>), download its sequences to our computer, and upload these sequences to the High-Performance Computing server (the IFB core cluster).</p></li>
<li><p>Download a sequence of interest (e.g. Histone H3) on the HPC server.</p></li>
<li><p>Use the similarity search tool <code>blast</code> to find homologs of the protein of interest in the selected genome.</p></li>
</ol>
<hr />
</div>
<div id="prerequisite" class="section level1">
<h1>Prerequisite</h1>
<p>This practical assumes that you are connected to a cluster of the IFB NNCR. This enables you to benefit from a software environment with pre-installed software tools for the different fields of bioinformatics.</p>
<p>You should have recieved a login before starting the course. If this is not the case, see the information on account request on the IFB Web site (<a href="https://www.france-bioinformatique.fr/en/cluster" class="uri">https://www.france-bioinformatique.fr/en/cluster</a>).</p>
<p>Before starting this tutorial, you should have followed the following introductory sessions</p>
<ul>
<li><p>Slides: <a href="https://ifb-elixirfr.gitlab.io/cluster/trainings/slurm/ebai2019.html">IFB High Performance Computing</a></p></li>
<li><p><a href="https://ifb-elixirfr.gitlab.io/cluster/doc/">IFB cluster doc</a></p></li>
<li><p><a href="https://ifb-elixirfr.gitlab.io/cluster/doc/quick-start/">Quick start</a></p></li>
<li><p><a href="https://ifb-elixirfr.gitlab.io/cluster/doc/slurm_user_guide/">Slurm user guide</a></p></li>
</ul>
</div>
<div id="warning-about-command-line-use-of-the-cluster" class="section level1">
<h1>Warning about command-line use of the cluster</h1>
<p>Please remember: <strong>the sequence analysis tasks below should by no way run on the login node of the computer</strong>. Each job has to be sent to the <code>slurm</code> job scheduler using <code>srun</code> (submit a single command) or <code>sbacth</code> (submit a bash script), as explained in the tutorials of the prerequisite section.</p>
<hr />
</div>
<div id="bioinformatics-resources-used-for-this-course" class="section level1">
<h1>Bioinformatics resources used for this course</h1>
<table>
<colgroup>
<col width="11%" />
<col width="56%" />
<col width="31%" />
</colgroup>
<thead>
<tr class="header">
<th>Resource</th>
<th>Description</th>
<th>URL</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Ensembl</strong></td>
<td>Genome database</td>
<td><a href="http://www.ensembl.org/">www.ensembl.org</a></td>
</tr>
<tr class="even">
<td><strong>UniprotKB</strong></td>
<td>Knowledge base of protein sequence and functional information</td>
<td><a href="https://www.uniprot.org/">www.uniprot.org</a></td>
</tr>
<tr class="odd">
<td><strong>PFAM</strong></td>
<td>Database of protein families</td>
<td><a href="https://pfam.xfam.org/">pfam.xfam.org</a></td>
</tr>
</tbody>
</table>
<hr />
</div>
<div id="collecting-the-sequences" class="section level1">
<h1>Collecting the sequences</h1>
<div id="loading-the-required-modules-on-the-ifb-core-cluster" class="section level2">
<h2>Loading the required modules on the IFB core cluster</h2>
<p>The IFB core cluster comes with a modular software environment, which enables users to deploy the specific tools required for their analyses.</p>
<p>In this section, we will open a connection to the IFB cluster and load the modules required for the practicals below.</p>
<p>Task 1: use the command <code>module avail</code>to explore the lists of available modules and identify the two modules used for this course.</p>
<ul>
<li><strong>blast</strong>: Basic Local Alignment Search Tool</li>
<li><strong>emboss</strong>: a software suite implemented by the European bioinformatics network EMBNet</li>
</ul>
<p>The solution can be seen by unfolding the code below.</p>
<p>Task 2: Load the required modules.</p>
</div>
<div id="creating-a-workspace-for-this-tutorial" class="section level2">
<h2>Creating a workspace for this tutorial</h2>
<p>In your home directory on the IFB core cluster, create a folder named <code>cmb-comp-bio_ptracticals</code>, and a sub-folder named <code>proteome-blast_results</code>.</p>
</div>
<div id="collecting-the-proteome-of-interest-from-uniprot" class="section level2">
<h2>Collecting the proteome of interest from Uniprot</h2>
<p>We will use Uniprot Web interface to select a protein of interest, that will serve as query to search for homologous proteins in the selected proteome.</p>
<p>For this tutorial, we will use the fruit fly <em>Drosophila melanogaster</em> as organism of interest</p>
<ul>
<li><p>Open a connection to <a href="http://uniprot.org">uniprot.org</a></p></li>
<li><p>Use the advanced serarches to formulate structured queries</p></li>
<li><p>Select the Organism. Start typing <em>Drosophila melanogaster</em> and wait for a few seconds. Among the possible completions displayed, select the following one.</p>
<p><em>Drosophila melanogaster (Fruit fly) [7227]</em></p>
<p>Note: the number <a href="https://www.ncbi.nlm.nih.gov/taxonomy/?term=7227">7227</a> is the identifier of this species in the NCBI taxonomic database.</p></li>
<li><p>Click <strong>Search</strong>. This will select all the proteins of the selected organism. The result table only displays a summary information for 25 entries.</p></li>
<li><p>On the top of the result page, click <strong>Download</strong> and download all proteins in fasta format on your computer.</p>
<p>Note: by default the sequence file is compressed with the <code>gzip</code> algorithm, you should thus obtain a binary file with the extension <code>.fasta.gz</code>.</p></li>
<li><p>Locate the downloaded file on your computer, and rename if with a simpler name, for example <code>Drosophila_melanogaster_proteome.fasta.gz</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p></li>
<li><p>Use the terminal to check the begining of this file</p></li>
</ul>
</div>
<div id="transferring-a-sequence-file-to-your-account-on-ifb-core-cluster" class="section level2">
<h2>Transferring a sequence file to your account on IFB core cluster</h2>
<div id="practical" class="section level3">
<h3>Practical</h3>
<ol style="list-style-type: decimal">
<li>Open a new terminal on your computer, and use the Unix command <code>scp</code> to copy the proteome sequence file in the result directory of the core cluster.</li>
</ol>
<p>Note that an alternative is to use a user-friendly program (e.g. Filezilla) to transfer this file. For more information read the section on <a href="https://ifb-elixirfr.gitlab.io/cluster/doc/data/">data transfer</a> in the IFB cluster documentation.</p>
<ol start="2" style="list-style-type: decimal">
<li><p>Come back to your <code>ssh</code> session on the IFB core cluster and check that the seque ce file is in the right directory, and contains the compressed proteome sequence file (extension <code>.fasta.gz</code>).</p></li>
<li><p>Measure the size (in megabase) of the compressed sequence file (command: <code>du</code>), then uncompress it (command: <code>gunzip</code>) and measure the size of the uncompressed file.</p></li>
</ol>
</div>
</div>
<div id="getting-a-query-sequence" class="section level2">
<h2>Getting a query sequence</h2>
<ul>
<li><p>Click on the Advanced search again. Your previous query is still displayed. Click the big <strong>+</strong> symbol to add a criterion, select <strong>Protein name</strong> and type <em>Histone H3</em>.</p></li>
<li><p>Click on the link to the protine named <em>Histone H3</em> (with ID <a href="https://www.uniprot.org/uniprot/P02299">P02299</a>). You will see the full record of annotations for this protein.</p></li>
<li><p>Select <strong>Format</strong> <span class="math inline">\(\rightarrow\)</span> <strong>Fasta</strong>. This displays the sequence of Histone 3 in fasta format. We will use this sequence as query for BLAST.</p></li>
<li><p>Copy the link of this sequence (<a href="https://www.uniprot.org/uniprot/P02299.fasta" class="uri">https://www.uniprot.org/uniprot/P02299.fasta</a>) in a file.</p></li>
<li><p>Come back to the IFB core cluster terminal, make sure you are in the results directory, and use the Unix command <code>wget</code> to download the sequence.</p></li>
</ul>
<p>Note: we use the option –no-clobber. To know why, read the wget manual after typing the command <code>man wget</code>. In Unix man pages, you can find a given word (e.g. clobber) by typing <code>/</code> followed by the search string.</p>
<hr />
</div>
</div>
<div id="searching-sequence-similarities-with-blast" class="section level1">
<h1>Searching sequence similarities with BLAST</h1>
<div id="formatting-the-blast-database-for-the-reference-genome" class="section level2">
<h2>Formatting the blast database for the reference genome</h2>
<p>Before being able to run sequence similarity searches, BLAST needs to read the sequence database (in our case, the proteome) in order to build an index of the k-mers (oligopeptides). This is done with the command <code>makeblastdb</code>.</p>
<div id="practical-1" class="section level3">
<h3>Practical</h3>
<ol style="list-style-type: decimal">
<li><p>Display the <code>makeblastdb</code> options (<code>-h</code>) manual (<code>-help</code>). Well, the the manual is quite heavy, so we give you the equired options: <code>-in</code> and <code>-dbtype</code>. Read the usage of these options in the manual, built the command and run it.</p></li>
<li><p>Build a command (<strong>DO NOT RUN IT YET</strong>) to index the proteome of interest.</p></li>
<li><p>Send this command to the job scheduler via <code>srun</code>.</p></li>
<li><p>Check the result files.</p></li>
</ol>
<p>The program <code>makeblastdb</code> created three files besides the input fasta file.</p>
<pre><code>Drosophila_melanogaster_proteome.fasta.pin
Drosophila_melanogaster_proteome.fasta.phr
Drosophila_melanogaster_proteome.fasta.psq</code></pre>
<p>These files contain an index of all the oligopeptites found in all the sequences of the fasta file. This “k-mer” index will be used by BLAST to rapidly find all the sequences in the database that match a query sequence, and use these short correspondences to start an alignment.</p>
</div>
</div>
<div id="read-blastp-manual" class="section level2">
<h2>Read <em>blastp</em> manual</h2>
<p>Since we want to search a protein database with protein query sequences, we will use the <code>blastp</code> command.</p>
<p>OK, the first contact is a bit frightening, because blastp has many options. This is because it allows you to run queries in different modes, with different parameters, and to export the results in different formats.</p>
<p>In this tutorial we will show you the most usual ways to use the tool, and you will then be able to refine your search by combing back to the manual and understanding the use of additional options.</p>
</div>
<div id="similaroty-search-for-one-sequence-against-a-database" class="section level2">
<h2>Similaroty search for one sequence against a database</h2>
<ol style="list-style-type: decimal">
<li><p>Build a command (<strong>DO NOT RUN IT YET</strong>) that searches similarities for one or several query sequences provided in a fasta file (e.g. <code>P02299.fasta</code>) in the proteome of interest (<code>Drosophila_melanogaster_proteome.fasta</code>). By default, <code>blastp</code> prints the result on the screen, but we prefer to redirect it to a file (e.g. <code>P02299_blastp_result.txt</code>) in order to keep a trace of the result and inspect it later.</p></li>
<li><p>Send this command to the job scheduler with <code>srun</code>.</p></li>
</ol>
<div id="questions" class="section level3">
<h3>Questions</h3>
<ol style="list-style-type: decimal">
<li>How many hits were found in total?</li>
<li>How many of these have an E-value higher than 1?</li>
<li>What was the threshold on the E-value for this blastp search?</li>
<li>With this threshold, how many hits would we expeect if we use a random sequence as query?</li>
<li>How many alignments would you consider significant in this result?</li>
<li>How many of these would you qualify of homologs?</li>
<li>For each of these putative homologs, is it a paralog, ortholog or “other log”?</li>
<li>Do you see a relationship between the E-value and some properties of the alignments?</li>
</ol>
</div>
</div>
<div id="getting-a-tabular-synthesis-of-blast-results" class="section level2">
<h2>Getting a tabular synthesis of blast results</h2>
<p>By default, BLAST displays the detailed results of a similaty search, starting with a summary of the matches, followed by all the alignemnts. It can be convenient to generate a synthetic result indicating only the relevant statistics for each alignment. This can be done by tuning BLAST options.</p>
<ol style="list-style-type: decimal">
<li><p>Read blastp manual and find a way to tune the formatting options in order to get a tabular output with comments.</p></li>
<li><p>Refine the query by selecting the following fields in the tabilar option:</p>
<ul>
<li>ID of the “subject” sequence (the one found in the proteome of interest)</li>
<li>length of the query sequence</li>
<li>length of the subject sequence</li>
<li>raw score of the alignment</li>
<li>e-value</li>
</ul></li>
</ol>
<div id="solution" class="section level3">
<h3>Solution</h3>
</div>
</div>
</div>
<div id="negative-control-blast-against-a-shuffled-sequence" class="section level1">
<h1>Negative control: blast against a shuffled sequence</h1>
<p>In this section, we will perform a quick empirical test, by submitting a random sequence to blastp. For this, we will shuffle the aminoacids of the original query sequence.</p>
<div id="shuffling-sequences-with-emboss-shuffleseq" class="section level2">
<h2>Shuffling sequences with EMBOSS shuffleseq</h2>
<ol style="list-style-type: decimal">
<li>Load the emboss module</li>
<li>Read the manual of the <code>shuffleseq</code> tool.</li>
<li>Send a command to the job scheduler to generate a shuffled version of the query sequence.</li>
</ol>
<p>Here is an example of result (I just simplified the header of the shuffle output fasta sequence for the sake of readability).</p>
<p>Of course, your result should differ since each shuffling produces a randomized sequence.</p>
<pre><code>&gt;shuffled_H3_DROME
LMKPEQMLIAKKKDTKRVLASVRGIHRIARKVARTERLQKVLFRKTIAFALAGDICGQPT
ESAPRTADLAQAVIFRHGTRPTRYTAPLSMQGKSYNLFGRQEDEQRASARASELTAIPKL
KVYERGAKAKQTRRLR</code></pre>
</div>
<div id="exercise" class="section level2">
<h2>Exercise</h2>
<ol style="list-style-type: decimal">
<li><p>Shuffle the histone 3 sequence with the tool <code>shuffleseq</code> (found in the conda environment <code>emboss</code>).</p></li>
<li><p>Run a <code>blastp</code> search with this shuffled sequence against the selected proteome.</p></li>
</ol>
<div id="solution-1" class="section level3">
<h3>Solution</h3>
</div>
<div id="questions-1" class="section level3">
<h3>Questions</h3>
<ol style="list-style-type: decimal">
<li>How many alignments would you expect by chance with the default blastp parameters?</li>
<li>How many alignemnts did you get with the shuffled sequence?</li>
<li>How many of these had an E-value smaller than 1?</li>
<li>How many of these had an E-value smaller than 0.01?</li>
<li>How many of these would you qualify of homologs?</li>
</ol>
<hr />
</div>
</div>
</div>
<div id="homework" class="section level1">
<h1>Homework</h1>
<hr />
</div>
<div id="supplementary-information-not-part-of-the-practical" class="section level1">
<h1>Supplementary information (not part of the practical)</h1>
<div id="installing-the-conda-environment-in-your-own-computer" class="section level2">
<h2>Installing the conda environment in your own computer</h2>
<p><strong>Attention</strong>, this is a complement to the practical, which should not be run if you are working on the IFB core cluster. This section explains you how to install on your environment all the tools required for this tutorial.</p>
<hr />
</div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>The file downloaded from Uniprot has a sophisticated name with parentheses and brackets, which will be hard to handle in unix commands: <code>uniprot-organism__Drosophila+melanogaster+(Fruit+fly)+[7227]_.fasta.gz</code><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
